sencha-workspace/SlateTasksTeacher/app/controller/Dashboard.js

```diff
@@ -458,8 +468,276 @@ Ext.define('SlateTasksTeacher.controller.Dashboard', {
+            Ext.Msg.alert('Error', 'Unable to authenticate user. Please try again or contact an administrator.');
+            return;
+        }
+
+        _userIsNotWithinDomain = function(errorMessage) {
+            Ext.Msg.alert('Error', errorMessage);
+            return;
+        };
+
+        _verifyUserIsWithinDomain = function(response) {
+            console.info('loaded user', response.result);
+
+            if (!response.result || !response.result.user || !response.result.user.emailAddress.match(googleAppsEmailRegex)) {
+                return _userIsNotWithinDomain('Account must be under the correct domain. ' + googleAppsDomain);
+                // return false;
+            }
+
+            tokenExpiration = new Date(authResult.expires_at * 1000);
+
+            Ext.util.Cookies.set('googleAppsToken', authResult.access_token, tokenExpiration, '/');
+            Ext.util.Cookies.set('googleAppsPermissionId', response.result.user.permissionId, tokenExpiration, '/');
+
+            me.doOpenFilePicker(authResult.access_token);
+        };
+
+        // confirm authentication is within google apps domain
+        gapi.client.request({
+            path: '/drive/v3/about',
+            params: {
+                fields: 'user'
+            }
+        }).
+        then(_verifyUserIsWithinDomain);
+    },
+
+    onPickerSelect: function(data) {
+        console.info('pickerCallback(%o)', data);
+        var me = this,
+            taskEditor = me.getTaskEditor(),
+
+            googleAppsDomain = SiteEnvironment && SiteEnvironment.googleAppsDomain,
+            userPermissionId = Ext.util.Cookies.get('googleAppsPermissionId', '/'),
+            googleAppsToken = Ext.util.Cookies.get('googleAppsToken', '/'),
+
+            showTaskEditor = [
+                google.picker.Action.CANCEL,
+                google.picker.Action.PICKED
+            ].indexOf(data[google.picker.Response.ACTION]) > -1,
+            filePicked = data[google.picker.Response.ACTION] == google.picker.Action.PICKED,
+
+            doc = filePicked && data[google.picker.Response.DOCUMENTS][0],
+
+            _getGoogleFileOwner = Ext.bind(me.getGoogleFileOwner, me),
+            _getGoogleFileOwnerEmail = function(response) {
+                return me.getGoogleFileOwnerEmail(response, doc);
+            },
+            _doVerifyGoogleFileOwner = function(response) {
+                return me.doVerifyGoogleFileOwner(response, doc);
+            };
+
+        if (showTaskEditor) {
+            taskEditor.show(true);
+        }
+
+        if (filePicked) {
+            gapi.client.request({
+                path: '/drive/v3/files/'+doc[google.picker.Document.ID]+'/permissions',
+                method: 'GET',
+                params: {
+                    'access_token': googleAppsToken
+                }
+            }).
+            then(_getGoogleFileOwner).
+            then(_getGoogleFileOwnerEmail).
+            then(_doVerifyGoogleFileOwner);
```

Comment:
@nbey again, inline functions if they're not being reused so we can see all the code in sequence